File: backend/app/services/advanced_conversation_state_service.py
import logging
from typing import Dict, Any, Optional
from datetime import datetime
from sqlalchemy.orm import Session
from sqlalchemy import and_

from ..models.conversation_state import ConversationState, ConversationStateType, IntentType
from ..schemas.conversation_state import StateTransitionRequest, ContextUpdateRequest, MemoryUpdateRequest

logger = logging.getLogger(__name__)


class AdvancedConversationStateService:
    def __init__(self, db: Session):
        self.db = db
    
    def create_or_get_state(self, session_id: str, user_id: int) -> ConversationState:
        """Create or get conversation state for a session"""
        try:
            state = self.db.query(ConversationState).filter(
                ConversationState.session_id == session_id
            ).first()
            
            if not state:
                state = ConversationState(
                    session_id=session_id,
                    user_id=user_id,
                    current_state="initial",
                    state_type=ConversationStateType.ACTIVE,
                    state_data={},
                    short_term_memory={},
                    long_term_memory={},
                    memory_priority={},
                    active_threads=["main"],
                    thread_priorities={"main": 1},
                    thread_contexts={"main": "general"}
                )
                self.db.add(state)
                self.db.commit()
                self.db.refresh(state)
            
            return state
        except Exception as e:
            logger.error(f"Error creating/getting state: {e}")
            self.db.rollback()
            raise
    
    def transition_state(self, request: StateTransitionRequest) -> ConversationState:
        """Transition conversation state"""
        try:
            state = self.db.query(ConversationState).filter(
                ConversationState.session_id == request.session_id
            ).first()
            
            if not state:
                raise ValueError(f"State not found for session {request.session_id}")
            
            # Update state
            state.previous_state = state.current_state
            state.current_state = request.to_state
            state.transition_reason = request.transition_reason
            state.transition_data = request.context_data
            state.last_activity = datetime.now()
            
            # Update context if provided
            if request.context_data:
                if "primary_context" in request.context_data:
                    state.primary_context = request.context_data["primary_context"]
                if "secondary_contexts" in request.context_data:
                    state.secondary_contexts = request.context_data["secondary_contexts"]
            
            self.db.commit()
            self.db.refresh(state)
            
            return state
            
        except Exception as e:
            logger.error(f"Error transitioning state: {e}")
            self.db.rollback()
            raise
    
    def update_context(self, request: ContextUpdateRequest) -> ConversationState:
        """Update conversation context"""
        try:
            state = self.db.query(ConversationState).filter(
                ConversationState.session_id == request.session_id
            ).first()
            
            if not state:
                raise ValueError(f"State not found for session {request.session_id}")
            
            # Update context fields
            if request.primary_context:
                state.primary_context = request.primary_context
            if request.secondary_contexts:
                state.secondary_contexts = request.secondary_contexts
            if request.state_data:
                state.state_data = request.state_data
            
            state.last_activity = datetime.now()
            self.db.commit()
            self.db.refresh(state)
            
            return state
            
        except Exception as e:
            logger.error(f"Error updating context: {e}")
            self.db.rollback()
            raise
    
    def update_memory(self, request: MemoryUpdateRequest) -> ConversationState:
        """Update conversation memory"""
        try:
            state = self.db.query(ConversationState).filter(
                ConversationState.session_id == request.session_id
            ).first()
            
            if not state:
                raise ValueError(f"State not found for session {request.session_id}")
            
            if request.memory_type == "short_term":
                if not state.short_term_memory:
                    state.short_term_memory = {}
                state.short_term_memory[request.memory_key] = request.memory_value
                
            elif request.memory_type == "long_term":
                if not state.long_term_memory:
                    state.long_term_memory = {}
                state.long_term_memory[request.memory_key] = request.memory_value
            
            # Update priority
            if not state.memory_priority:
                state.memory_priority = {}
            state.memory_priority[request.memory_key] = request.priority
            
            state.last_activity = datetime.now()
            self.db.commit()
            self.db.refresh(state)
            
            return state
            
        except Exception as e:
            logger.error(f"Error updating memory: {e}")
            self.db.rollback()
            raise
    
    def get_complex_state(self, session_id: str, user_id: int) -> Dict[str, Any]:
        """Get complex conversation state with all related data"""
        try:
            state = self.db.query(ConversationState).filter(
                ConversationState.session_id == session_id
            ).first()
            
            if not state:
                raise ValueError(f"State not found for session {session_id}")
            
            return {
                "session_id": state.session_id,
                "current_state": state.current_state,
                "previous_state": state.previous_state,
                "state_type": state.state_type,
                "primary_context": state.primary_context,
                "secondary_contexts": state.secondary_contexts,
                "state_data": state.state_data,
                "short_term_memory": state.short_term_memory,
                "long_term_memory": state.long_term_memory,
                "memory_priority": state.memory_priority,
                "active_threads": state.active_threads,
                "thread_priorities": state.thread_priorities,
                "thread_contexts": state.thread_contexts,
                "last_activity": state.last_activity,
                "created_at": state.created_at
            }
            
        except Exception as e:
            logger.error(f"Error getting complex state: {e}")
            raise
    
    def track_conversation_event(self, session_id: str, user_id: int, event_type: str, event_data: Dict[str, Any]) -> Dict[str, Any]:
        """Track conversation event"""
        try:
            # Create or get state
            state = self.create_or_get_state(session_id, user_id)
            
            # Add event to short-term memory
            memory_key = f"event_{event_type}_{datetime.now().timestamp()}"
            memory_value = {
                "type": event_type,
                "data": event_data,
                "timestamp": datetime.now().isoformat()
            }
            
            if not state.short_term_memory:
                state.short_term_memory = {}
            state.short_term_memory[memory_key] = memory_value
            
            # Update memory priority
            if not state.memory_priority:
                state.memory_priority = {}
            state.memory_priority[memory_key] = 1
            
            state.last_activity = datetime.now()
            self.db.commit()
            self.db.refresh(state)
            
            return {"id": 1, "session_id": session_id, "event_type": event_type}
            
        except Exception as e:
            logger.error(f"Error tracking conversation event: {e}")
            self.db.rollback()
            raise

File: src/services/advancedConversationStateService.ts
import { apiRequest } from './api';

// Interfaces para gestión avanzada de estado de conversaciones
export interface ConversationState {
  id: number;
  session_id: string;
  user_id: number;
  current_state: string;
  state_type: 'active' | 'paused' | 'completed' | 'abandoned' | 'transitioning';
  state_data: Record<string, any>;
  primary_context?: string;
  secondary_contexts: string[];
  context_stack: Record<string, any>[];
  current_intent?: string;
  intent_type: 'primary' | 'secondary' | 'contextual' | 'fallback';
  intent_confidence: number;
  intent_history: Record<string, any>[];
  flow_id?: string;
  flow_step: number;
  flow_data: Record<string, any>;
  flow_history: Record<string, any>[];
  short_term_memory: Record<string, any>;
  long_term_memory: Record<string, any>;
  memory_priority: Record<string, number>;
  active_threads: string[];
  thread_priorities: Record<string, number>;
  thread_contexts: Record<string, any>;
  previous_state?: string;
  transition_reason?: string;
  transition_data: Record<string, any>;
  created_at: string;
  updated_at?: string;
  last_activity: string;
}

export interface StateTransitionRequest {
  session_id: string;
  to_state: string;
  transition_type: string;
  transition_reason?: string;
  context_data?: Record<string, any>;
  intent_data?: Record<string, any>;
}

export interface ContextUpdateRequest {
  session_id: string;
  primary_context?: string;
  secondary_contexts?: string[];
  context_stack?: Record<string, any>[];
}

export interface MemoryUpdateRequest {
  session_id: string;
  memory_type: 'short_term' | 'long_term';
  memory_key: string;
  memory_value: any;
  priority?: number;
}

export interface ComplexConversationState {
  session_id: string;
  current_state: string;
  contexts: Record<string, any>;
  intents: Record<string, any>[];
  flows: Record<string, any>[];
  threads: Record<string, any>[];
  memories: Record<string, any>;
  last_activity: string;
}

export class AdvancedConversationStateService {
  // Gestión de estado
  static async createOrGetState(sessionId: string): Promise<ConversationState> {
    const response = await apiRequest<ConversationState>('/conversation-state/state', {
      method: 'POST',
      body: { session_id: sessionId }
    });
    return response.data;
  }

  static async transitionState(request: StateTransitionRequest): Promise<ConversationState> {
    const response = await apiRequest<ConversationState>('/conversation-state/transition', {
      method: 'POST',
      body: request
    });
    return response.data;
  }

  // Gestión de contexto
  static async updateContext(request: ContextUpdateRequest): Promise<ConversationState> {
    const response = await apiRequest<ConversationState>('/conversation-state/context', {
      method: 'PUT',
      body: request
    });
    return response.data;
  }

  // Gestión de memoria
  static async updateMemory(request: MemoryUpdateRequest): Promise<ConversationState> {
    const response = await apiRequest<ConversationState>('/conversation-state/memory', {
      method: 'PUT',
      body: request
    });
    return response.data;
  }

  // Estado complejo
  static async getComplexState(sessionId: string): Promise<ComplexConversationState> {
    const response = await apiRequest<ComplexConversationState>(`/conversation-state/complex-state/${sessionId}`, {
      method: 'GET'
    });
    return response.data;
  }

  // Tracking de eventos
  static async trackConversationEvent(
    sessionId: string, 
    eventType: string, 
    eventData: Record<string, any>
  ): Promise<{ session_id: string; event_type: string; state_updated: boolean }> {
    const response = await apiRequest<{ session_id: string; event_type: string; state_updated: boolean }>('/conversation-state/track-conversation-event', {
      method: 'POST',
      body: { session_id: sessionId, event_type: eventType, event_data: eventData }
    });
    return response.data;
  }

  // Métodos de utilidad para el chatbot
  static async initializeConversationSession(sessionId: string): Promise<ConversationState> {
    return this.createOrGetState(sessionId);
  }

  static async updateConversationIntent(
    sessionId: string, 
    intent: string, 
    confidence: number = 0.8,
    intentType: 'primary' | 'secondary' | 'contextual' | 'fallback' = 'primary'
  ): Promise<ConversationState> {
    return this.transitionState({
      session_id: sessionId,
      to_state: 'intent_identified',
      transition_type: 'automatic',
      transition_reason: `Intent identified: ${intent}`,
      intent_data: {
        current_intent: intent,
        intent_type: intentType,
        intent_confidence: confidence
      }
    });
  }

  static async storeConversationMemory(
    sessionId: string,
    memoryType: 'short_term' | 'long_term',
    key: string,
    value: any,
    priority: number = 1
  ): Promise<ConversationState> {
    return this.updateMemory({
      session_id: sessionId,
      memory_type: memoryType,
      memory_key: key,
      memory_value: value,
      priority
    });
  }

  static async getConversationContext(sessionId: string): Promise<ComplexConversationState> {
    return this.getComplexState(sessionId);
  }
}


